### ISSUE TITLE
Implement Grid Cells for Compositional Generalization

### ISSUE THREAD

**Issue by DrDrewCain** [DrDrewCain/MiniMacLLM#24]
## Overview
Implement grid cell-inspired representations for compositional generalization and spatial/relational reasoning.

## Priority
⭐⭐⭐ (Medium-High Priority)

## Expected Benefits
- **Compositional generalization**: Systematic combination of learned concepts
- **Relational memory**: Similar experiences stored nearby in metric space
- **Zero-shot reasoning**: Path integration enables unseen combinations
- **Domain separation**: Different domains occupy different grid regions

## Biological Background

Grid cells in the entorhinal cortex fire in hexagonal patterns, creating a universal metric coordinate system for navigation. They enable:
- Path integration (tracking position from velocity)
- Spatial memory and relational reasoning
- Compositional representation of abstract spaces

### 1. Core Mathematics

**Neural Manifold as 2D Embedding**:
```python
v(x+Δx) = F(v(x), Δx)  # Forms Abelian Lie group
```

**Conformal Normalization** (key for hexagonal emergence):
```python
Δr̄ = s·Δr / ||f(v(x), θ)||
# Ensures distance-preserving position embeddings
```

**Loss Function**:
```python
L = L_0 + λ_1·L_1
# L_0: Basis expansion (Gaussian place cells)
# L_1: Transformation accuracy
```

### 2. Architecture Options

**Option 1: Positional Encoding Replacement**
- Replace RoPE with multi-scale grid cell representations
- 4-8 grid modules with different scales (0.3, 0.4, 0.5, 0.6...)

**Option 2: Knowledge Space Navigation**
- Navigate abstract knowledge space like physical space
- Estimate "velocity" in knowledge space
- Path integration for reasoning chains

**Option 3: Domain-Specific Coordinates**
- Each domain gets its own grid coordinate system
- Similar experiences retrieve together
- Natural domain separation

### 3. Implementation Files

Create:
- `src/neurobio/grid_cells.py` - GridCellModule, conformal normalization
- `src/model/grid_reasoning.py` - Optional reasoning module
- `tests/unit/test_grid_cells.py` - Unit tests
- `tests/integration/test_grid_integration.py` - Integration tests

### 4. Configuration Options

Add to `ContinualLearningConfig`:
```python
use_grid_cells: bool = False  # Optional module
grid_num_modules: int = 8  # Multi-scale
grid_scales: list = [0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0]
grid_use_for_position: bool = False  # Replace RoPE?
grid_use_for_reasoning: bool = True  # Reasoning module
```

## Key Research Papers

1. **Conformal Normalization**: Ben-Yosef & Derdikman, 2023. "Grid Cells and Conformal Geometry" arXiv:2310.19192
   - 82.5-87.6% neurons develop grid-like firing
   - Hexagonal patterns emerge naturally
   - Theoretical proof via Fourier analysis

2. **Conformal Isometry**: Ben-Yosef & Derdikman, 2024. "The Conformal Isometry Hypothesis" arXiv:2405.16865
   - Explains WHY hexagons (maximally distance-preserving)
   - Hexagon flat torus geometry

3. **Grid Cells in Cognition**: Bellmund et al., 2024. "Grid cells for conceptual spaces" Annual Reviews
   - Evidence for grid-like codes in abstract reasoning
   - Supports transitive inference

4. **Original Grid Cell Discovery**: Hafting et al., 2005. "Microstructure of a spatial map in the entorhinal cortex" Nature 436:801-806

5. **Computational Grid Cells**: Cueva & Wei, 2018. "Emergence of grid-like representations by training recurrent neural networks" ICLR 2018

### Phase 1: Core Grid Cell Module
1. Implement GridCellRNN with conformal normalization
2. Train on position prediction tasks
3. Verify hexagonal pattern emergence

### Phase 2: Integration Options
1. Test as positional encoding
2. Test as reasoning module
3. Benchmark on compositional tasks

### Phase 3: Domain-Specific Grids
1. Implement per-domain coordinate systems
2. Test retrieval similarity
3. Benchmark on continual learning

## Testing Requirements

1. **Unit Tests**:
   - Conformal normalization correctness
   - Grid pattern emergence
   - Multi-scale representations
   - RNN dynamics

2. **Integration Tests**:
   - Grid cells as position encoding
   - Knowledge space navigation
   - Domain separation
   - Compositional reasoning tasks

3. **Benchmarks**:
   - Zero-shot generalization
   - Compositional reasoning accuracy
   - Domain interference
   - Memory retrieval similarity

## Success Criteria

- [ ] GridCellModule with conformal normalization
- [ ] 80%+ neurons develop grid-like firing
- [ ] Multi-scale representations working
- [ ] Integration with LLM tested
- [ ] Compositional reasoning improved
- [ ] All tests passing

## Notes

- Grid cells are most useful for **specialized reasoning modules**, not core language modeling
- Consider making this an **optional enhancement** for specific task types
- May be more valuable for multi-hop QA, knowledge graphs, spatial reasoning

## References

See research summary in conversation context for detailed paper analysis.
